src/crimson/osd/pg.cc

```diff
@@ -890,11 +904,15 @@ PG::do_osd_ops_execute(
         });
       });
     }), OpsExecuter::osd_op_errorator::all_same_way(
-        [this, rollbacker, failure_func_ptr]
+        [this, error_func_ptr, rollbacker, failure_func_ptr]
         (const std::error_code& e) mutable {
           return rollbacker.rollback_obc_if_modified(e).then_interruptible(
-          [this, e, failure_func_ptr] {
-            return (*failure_func_ptr)(e , shard_services.get_tid());
+          [this, e, failure_func_ptr, error_func_ptr] {
+            auto rep_tid = shard_services.get_tid();
+            return (*error_func_ptr)(e, rep_tid)
+            .then([failure_func_ptr, e, rep_tid] {
+              return (*failure_func_ptr)(e , rep_tid);
+            });
```

Comment:
2/2: The second `submit_error_log` call, since it's being run from within `do_osd_ops_execute` it should also be safe . Even though it's packed inside `all_completed_fut` below (in case of an error), it will be executed before returning `pg_rep_op_fut_t` to the caller.

Reply:
Hmm, this handler is actually chained after _all_complete_fut and could therefore happen after we leave the exclusive phase.  On the other hand, I think this case is actually impossible -- any error here would have to come from the _all_completed_fut returned from submit_transaction, which would have to be an error with the actual transaction submitted to the backing store or somehow with the messages received from peers.  I *think* any error here could be treated as fatal to the OSD.

Reply:
> Hmm, this handler is actually chained after _all_complete_fut and could therefore happen after we leave the exclusive phase. On the other hand, I think this case is actually impossible -- any error here would have to come from the _all_completed_fut returned from submit_transaction, which would have to be an error with the actual transaction submitted to the backing store or somehow with the messages received from peers. I _think_ any error here could be treated as fatal to the OSD.

This case was possible since there were non-fatal errors that could have been returned as well. I added an assert to verify that we handle only those.
Moreover, I moved the call to error_log_ptr earlier - before actually returning the non-fatal errors. That way, it will be executed during the exclusive phase.